### PR TITLE
Ports Goon's chat window

### PR DESCRIPTION
**Pull request by lzimann**
Ports VG pr that ported goon's chat.

This replaces the chat window with a custom chat, giving more options to play around with.

I am going to commit the refactor of `target << message` to the new `to_chat()` proc after I'm done with them.

- [x] Change the css file so it matches ours instead of vg's

- [x] Finish `<<` operator replacement

- [x] Fix links not working(and some are not displaying properly)

- [x] Fix emojis not displaying like they should

- [ ] Figure out why messages have extra whitespaces on them.

chat

Some regexes I used to convert: 
`^\s*(?!diary|diaryofmeanpeople|world.log|log|S|F|Banlist|secret_satchels|runtime_diary|file)([^\s\d]\S*)\s*<<\s*(?!')([^\d(\s][^(]*)$`

`\\icon\[(.*)\]`

Just be aware that the regexes have some issues and I had to manually fix a LOT of `target << thing` instances.

### CODE REVIEW

**Comment by GunHog** [comment]
Pics plz k thx lol

**Comment by PKPenguin321** [comment]
>I am going to commit the refactor of target << message to the new to_chat() proc after I'm done with them.

Godspeed

File: goon/browserassets/css/font-awesome.css [hunk 1]
```diff
@@ -0,0 +1,788 @@
+@font-face{font-family:'FontAwesome';src:url('fontawesome-webfont.eot');src:url('fontawesome-webfont.eot') format('embedded-opentype'),url('fontawesome-webfont.woff') format('woff'),url('fontawesome-webfont.ttf') format('truetype'),url('fontawesome-webfont.svg') format('svg');font-weight:normal;font-style:normal;}[class^="icon-"],[class*=" icon-"]{font-family:FontAwesome;font-weight:normal;font-style:normal;text-decoration:inherit;-webkit-font-smoothing:antialiased;*margin-right:.3em;}
```
Comment (JJRcop) [inline goon/browserassets/css/font-awesome.css]: We already have fontawesome, though. Do we need two of them?

Comment (lzimann) [inline goon/browserassets/css/font-awesome.css]: Do we? I can use ours then

**Comment by Cyberboss** [comment]
>Gooncode

**Comment by Cyberboss** [comment]
Also how the hell do you plan on refactoring listener << message? Absolute madman

File: goon/browserassets/css/font-awesome.css (hunk 1 above)
Comment (JJRcop) [inline goon/browserassets/css/font-awesome.css]: Yeah, it is in the tgui/assets folder though, maybe it should be moved to a subdirectory of html

Comment (lzimann) [inline goon/browserassets/css/font-awesome.css]: I'll do that after I'm done with all this conversion

**Comment by Lancer411** [comment]
Bicon() kills animated images in chat from dmi by converting it to base64 with undocumented byond feature. This includes several emojies.

**Comment by Quardbreak** [comment]
Oh shit, this will broke "#&255;" and "&#1103;" characters again? :(

**Comment by RemieRichards** [comment]
@Quardbreak actually VG's version of this chat window can handle UTF8 characters by avoiding byond's internal decoding of it, and allowing the window itself to do it.

If anything those kinds of characters/letters will be BETTER supported.

File: goon/browserassets/html/browserOutput.html [hunk 2]
```diff
@@ -0,0 +1,44 @@
+<!DOCTYPE html>
+<html>
+<head>
+	<title>Chat</title>
+	<meta http-equiv="X-UA-Compatible" content="IE=edge,chrome=1" />
+	<meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
+	<link rel="stylesheet" type="text/css" href="font-awesome.css" />
+	<link rel="stylesheet" type="text/css" href="browserOutput.css" />
+	<script type="text/javascript" src="jquery.min.js"></script>
```
Comment (AnturK) [inline goon/browserassets/html/browserOutput.html]: Is this another version of jquery, or one of our current 5. These really should be trimmed to one.

File: goon/browserassets/js/browserOutput.js [hunk 3]
```diff
@@ -0,0 +1,944 @@
...
+function toHex(n) {
+	n = parseInt(n,10);
+	if (isNaN(n)) return "00";
+	n = Math.max(0,Math.min(n,255));
+	return "0123456789ABCDEF".charAt((n-n%16)/16) + "0123456789ABCDEF".charAt(n%16);
+}
+
+function changeMode(mode) {
+	switch (mode) {
+		case 'geocities':
```
Comment (AnturK) [inline goon/browserassets/js/browserOutput.js]: what

File: goon/code/datums/browserOutput.dm [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+
+	cookieSent = 1
+
+//Called by js client every 60 seconds
+/datum/chatOutput/proc/ping()
+	return "pong"
+
+//Called by js client on js error
+/datum/chatOutput/proc/debug(error)
+	log_world("\[[time2text(world.realtime, "YYYY-MM-DD hh:mm:ss")]\] Client: [(src.owner.key ? src.owner.key : src.owner)] triggered JS error: [error]")
```
Comment (AnturK) [inline goon/code/datums/browserOutput.dm]: This seems spammable

File: goon/code/datums/browserOutput.dm [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+
+/client/verb/debug_chat()
```
Comment (AnturK) [inline goon/code/datums/browserOutput.dm]: This should probably be in debug mode only.

File: goon/code/datums/browserOutput.dm [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+			message = replacetext(message, "\proper", "")
+
+		//Grab us a client if possible
+		var/client/C
+		if (istype(target, /client))
+			C = target
+		else if (istype(target, /mob))
+			C = target:client
+		else if (istype(target, /datum/mind) && target:current)
+			C = target:current:client
```
Comment (AnturK) [inline goon/code/datums/browserOutput.dm]: A single proc to get client/output target on /datum would be faster that all these istypes and :'s ?

**Comment by Quardbreak** [comment]
@RemieRichards I already tested Paradise Goonchat, there Cyrillic can possible work, but still need some sanitize changes to remove/replace those "я" characters.
But... If here gonna add better support for this - everything very good.

**Review by AnturK — COMMENTED** [review]
Just few things from quick glance

**Comment by RemieRichards** [comment]
@Quardbreak I don't believe paradise ported the update that makes it work nicely yet.

**Comment by xxalpha** [comment]
Pictures.

**Comment by Earthykiller127** [comment]
Steals*

File: goon/code/datums/browserOutput.dm [hunk 7]
```diff
@@ -0,0 +1,283 @@
...
+
+		log_world("Sending main chat window to client [owner.ckey]")
+		owner << browse(file("goon/browserassets/html/browserOutput.html"), "window=browseroutput")
+		sleep(20)
+		if(!owner || loaded)
+			break
+
+	// world.log << "chatOutput: load() completed"
+
+/datum/chatOutput/Topic(var/href, var/list/href_list)
```
Comment (RemieRichards) [inline goon/code/datums/browserOutput.dm]: remove all these ```var/``` from the entire PR.

File: goon/code/datums/browserOutput.dm [hunk 8]
```diff
@@ -0,0 +1,283 @@
...
+	if(usr.client != owner)
+		return 1
```
Comment (RemieRichards) [inline goon/code/datums/browserOutput.dm]: all of these to the proper ```TRUE/FALSE```

File: goon/code/datums/browserOutput.dm [hunk 9]
```diff
@@ -0,0 +1,283 @@
...
+	src.sendClientData()
+
+	pingLoop()
+
+/datum/chatOutput/proc/pingLoop()
+	set waitfor = FALSE
+
+	while (owner)
+		ehjax_send(data = owner.is_afk(29) ? "softPang" : "pang") // SoftPang isn't handled anywhere but it'll always reset the opts.lastPang.
+		sleep(30)
```
Comment (RemieRichards) [inline goon/code/datums/browserOutput.dm]: I'm sure we can handle this nicer.

@MrStonedOne @Cyberboss

File: goon/code/datums/browserOutput.dm (hunk 7 above)
Comment (lzimann) [inline goon/code/datums/browserOutput.dm]: Already on it.

File: goon/code/datums/browserOutput.dm [hunk 10]
```diff
@@ -0,0 +1,283 @@
...
+/datum/chatOutput/proc/analyzeClientData(cookie = "")
+	if(!cookie)
+		return
+
+	if(cookie != "none")
+		var/list/connData = json_decode(cookie)
+		if (connData && islist(connData) && connData.len > 0 && connData["connData"])
+			src.connectionHistory = connData["connData"] //lol fuck
+			var/list/found = new()
+			for (var/i = src.connectionHistory.len; i >= 1; i--)
```
Comment (RemieRichards) [inline goon/code/datums/browserOutput.dm]: for(var/i in connectionHistory.len to 0 step -1)

File: goon/code/datums/browserOutput.dm (hunk 9 above)
Comment (lzimann) [inline goon/code/datums/browserOutput.dm]: Reminder for myself: change the ping to use ours intead of this one.

File: goon/code/datums/browserOutput.dm [hunk 11]
```diff
@@ -0,0 +1,283 @@
...
+		var/client/C
+		if (istype(target, /client))
+			C = target
+		else if (istype(target, /mob))
+			C = target:client
+		else if (istype(target, /datum/mind) && target:current)
+			C = target:current:client
+
+		if (C && C.chatOutput)
+			if(C.chatOutput.broken) // Either a secret repo fuck up or a player who hasn't updated his skin file.
```
Comment (RemieRichards) [inline goon/code/datums/browserOutput.dm]: remove mention of a secret repo, that's a vg thing.

Comment (lzimann) [inline goon/code/datums/browserOutput.dm]: Already did.

**Comment by Earthykiller127** [comment]
Steals*

File: goon/browserassets/js/browserOutput.js [hunk 12]
```diff
@@ -0,0 +1,944 @@
...
+			opts.chatMode = mode;
+			break;
+		case 'default':
+		default:
+			//remove loaded stylesheet/s
+			opts.chatMode = 'default';
+	}
+}
+
+function handleClientData(ckey, ip, compid) {
```
Comment (Chnkr) [inline goon/browserassets/js/browserOutput.js]: If you guys end up porting this specific function as well (vg did not) make sure you at least change the following

- `?_src_=chat&proc=analyzeClientData` will show up in the href logs, make sure you actually remove the cid/ip from the public logs, as /vg/ did not which is bad

**Comment by MrPerson** [comment]
where are the pictures egg man you said there would be pictures

**Comment by lzimann** [comment]
Updated OP with one pic of the chat

**Comment by Chnkr** [comment]
@Quardbreak you might want to take a look at this https://github.com/d3athrow/vgstation13/pull/13537

**Comment by rainbowjose** [comment]
Mother of god. That will breaks my russian letter santizes completely.

File: goon/code/datums/browserOutput.dm (hunk 9 above)
Comment (Cyberboss) [inline goon/code/datums/browserOutput.dm]: Why is this even necessary?

**Comment by Injazz** [comment]
@lzimann PJB did some kind of Python magic to switch the << easily. Just ask him to give that script.

**Comment by Injazz** [comment]
@Quardbreak
@rainbowjose 
The only thing you need is just to wait when the << refactor happens.
After that everything you need is just set the letter replacement in the _to_chat proc
REMEMBER: since Goonchat is the HTML magic thingy, you need to replace problematic letter with the `&#1103;` (html encode) instead of `&#255;`
https://github.com/Signmark/ATMTA-Station/commit/4d3080b61bcb459a35e1561c08e4d68588691d15
https://github.com/Signmark/ATMTA-Station/commit/d34c9cc27c7139fb9b5f36cb5d1a2fffc5801985
see this for reference

**Comment by Hatfish** [comment]
There isn't a way to hide the ping indicator and settings window, is there?

**Comment by lzimann** [comment]
The ping yes, settings no(should be trivial to do if people really want though).

File: goon/browserassets/js/browserOutput.js (hunk 12 above)
Comment (MrStonedOne) [inline goon/browserassets/js/browserOutput.js]: Our href logs are not public, would it show in any other log?

Comment (Chnkr) [inline goon/browserassets/js/browserOutput.js]: Not that I'm aware of, looks fine then 👍

File: goon/browserassets/js/browserOutput.js (hunk 3 above)
Comment (Supermichael777) [inline goon/browserassets/js/browserOutput.js]: now its a font shitting up your browser

File: code/modules/emoji/emoji_parse.dm [hunk 13]
```diff
@@ -19,7 +19,7 @@ var/list/emojis
 			if(search)
 				emoji = lowertext(copytext(text, pos+1, search))
 				if(emoji in emojis)
-					parsed += " <img class=icon src=\ref['icons/emoji.dmi'] iconstate='[emoji]'>"
+					parsed += bicon(icon('icons/emoji.dmi', emoji))
```
Comment (Lancer411) [inline code/modules/emoji/emoji_parse.dm]: Check it with animated emojies. Bicon converts them to png instead of gif.

Comment (lzimann) [inline code/modules/emoji/emoji_parse.dm]: We don't have animated emojis here.

Comment (Lancer411) [inline code/modules/emoji/emoji_parse.dm]: last time i checked ( it was two weeks ago ) there was.
:supermatter: and :forcesword:
It also breaks door icons on examine on our build.

Comment (JJRcop) [inline code/modules/emoji/emoji_parse.dm]: It appears to take all the frames and throw them into a grid


Best solution is to ban animated emoji and remove the extra frames from the emoji.dmi, which are supermatter and forcesword

Comment (lzimann) [inline code/modules/emoji/emoji_parse.dm]: I'll do that. After this gets merged, someone can look into making all the emotes client-sided(which is possible with the browser) and allow animated emotes and what not.

Comment (Lancer411) [inline code/modules/emoji/emoji_parse.dm]: In the end i used js and base64 css file for client-side emoticons, but examining of animated icons is sill an issue.

**Comment by PKPenguin321** [comment]
So just curious, what kind of new features will this allow aside from the nasty ping display thing

**Comment by lzimann** [comment]
We could make radio colors a chat preference instead of having a war every 2 months when someone tries to change it. We can make emotes client-sided(instead of server sided which is now). And probably much more that I am not aware.

**Comment by Iamgoofball** [comment]
We can embed HTML in the chat window

**Comment by Razharas** [comment]
We can make different radio chants different tabs

**Comment by MrStonedOne** [comment]
We could have a special right click menu for certain types of words (like replace all the admin links with right click options) Something that would also help with the identity system

**Comment by Core0verload** [comment]
We could finally fix the #&255; forever.

Can't wait for this to get done. Damn, I'm ready to replace << manually for it to happen.

**Comment by lzimann** [comment]
I am still having issues with links not applying the visited state and some icons displaying more than one state when you examine them. Other than that this is pretty much ready.

**Comment by maxyo** [comment]
@lzimann Hey dude. I'm porting this shit too. Can you tell me full regex code to replace all \icon to bicon, please?

**Comment by lzimann** [comment]
@maxyo I used the one I posted in the OP, but it was not working properly and I pretty much had to fix them all manually(it has some grouping errors).

**Comment by maxyo** [comment]
@lzimann 
>Figure out why messages have extra whitespaces on them.

You mean this shit:
http://prntscr.com/ec7734 ?

**Comment by lzimann** [comment]
@maxyo Exactly, but for some reason it stopped happening for me/when I testmerged it in the live server, so I have absolutely no idea why it is happening.

**Comment by Mervill** [comment]
I've been out of the loop recently but I'm, %100 on board with this!

**Comment by maxyo** [comment]
@lzimann I know why it happening. When you add some additional words to that fields http://prntscr.com/ecds8b

**Comment by optimumtact** [comment]
ping to reopen ormake a new pr when you're ready

**Comment by lzimann** [comment]
I'll see if I can figure out the links not applying visited state and icons showing more than one state on examine... but no promises.

**Comment by Core0verload** [comment]
NOOOOOOOOOO

**Comment by rainbowjose** [comment]
Baby, its triple

**Pull request closed without merging**